# 데이터 사전처리 (Preprocessing)
- 머신러닝 등 데이터 분석의 정확도는 분석 데이터의 품질에 좌우된다.
- 데이터의 품질을 높이기 위해서는 누락 데이터, 중복 데이터 등 오류를 수정하고 분석목적에 맞게 변형하는 과정이 필요
- 수집한 데이터를 분석 목적에 적합하도록 하는 작업을 사전 처리(Preprocessing)이라고 한다. 

### 누락 데이터(NaN : Not a Number)
- 데이터를 파일로 입력할 때 빠트리거나 파일 형식을 변한하는 과정에서 손실되는 경우가 발생
- 머신러닝 분석 모형에 데이터를 입력하기 전에 반드시 누락 데이터를 제거하고 다른 적절한 값으로 대체 하여야 한다. 
- 누락 데이터가 많아지면 데이터의 품질이 떨어지고 머신러닝 분석 알고리즘을 왜곡하는 현상이 발생한다.

In [397]:
import seaborn as sns

In [398]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [399]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


> deck에는 203개의 유효한 번주형 데이터만 있고 나머지 688개가 NaN 이다. 

In [400]:
# deck열의 NaN갯수 계산하기
df.deck.value_counts(dropna=False)

NaN    688
C       59
B       47
D       33
E       32
A       15
F       13
G        4
Name: deck, dtype: int64

### 누락 데이터 찾기

In [401]:
# NaN 찾기 
df.head().isnull()

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0     False   False  False  False  False  False  False     False  False   
1     False   False  False  False  False  False  False     False  False   
2     False   False  False  False  False  False  False     False  False   
3     False   False  False  False  False  False  False     False  False   
4     False   False  False  False  False  False  False     False  False   

     who  adult_male   deck  embark_town  alive  alone  
0  False       False   True        False  False  False  
1  False       False  False        False  False  False  
2  False       False   True        False  False  False  
3  False       False  False        False  False  False  
4  False       False   True        False  False  False

In [402]:
# NaN이 아닌 것 찾기
df.head().notnull()

survived  pclass   sex   age  sibsp  parch  fare  embarked  class   who  \
0      True    True  True  True   True   True  True      True   True  True   
1      True    True  True  True   True   True  True      True   True  True   
2      True    True  True  True   True   True  True      True   True  True   
3      True    True  True  True   True   True  True      True   True  True   
4      True    True  True  True   True   True  True      True   True  True   

   adult_male   deck  embark_town  alive  alone  
0        True  False         True   True   True  
1        True   True         True   True   True  
2        True  False         True   True   True  
3        True   True         True   True   True  
4        True  False         True   True   True

In [403]:
# 누락 데이터 갯수 확인
df.isnull().sum(axis = 0)

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [404]:
# 누락 데이터 제거 
missing_df = df.isnull()


In [405]:
for col in missing_df.columns:
    missing_count = (missing_df[col].value_counts())
    try:
        print("%12s"%col, ":", missing_count[True])
    except:
        print("%12s"%col,":",0)
    



    survived : 0
      pclass : 0
         sex : 0
         age : 177
       sibsp : 0
       parch : 0
        fare : 0
    embarked : 2
       class : 0
         who : 0
  adult_male : 0
        deck : 688
 embark_town : 2
       alive : 0
       alone : 0


In [406]:
# NaN값이 500개 이상인 열을 모두 삭제
df_thresh = df.dropna(axis=1, thresh= 500)
df_thresh.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'embark_town', 'alive',
       'alone'],
      dtype='object')

In [407]:
# age 열에 나이 데이터가 없는 모든 행을 삭제
df_thresh_age = df_thresh.dropna(subset=['age'],how='any',axis=0)
len(df_thresh_age)

714

---
### 누락 데이터의 치환
- 누락 데이터를 무작정 삭제해 버리면 어렵게 수집한 데이터를 활용하지 못한다.
- 데이터 중 일부가 누락되어 있더라도 나머지 데이터를 최대한 살려서 데이터 분석에 활용하는 경우가 좋은 결과를 얻는 경우가 많다.

In [408]:
df = sns.load_dataset('titanic')

In [409]:
# age 열의 첫 10개 데이터 출력

df.age.head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: age, dtype: float64

In [410]:
# age 열의 NaN값을 다른 나이 데이터의 평균으로 변경하기
# df['age'][df.age.isnull()] = df.age[df.age.notnull()].mean()
df.age.fillna(df.age.mean(),inplace=True)
df.age.head(10)

0    22.000000
1    38.000000
2    26.000000
3    35.000000
4    35.000000
5    29.699118
6    54.000000
7     2.000000
8    27.000000
9    14.000000
Name: age, dtype: float64

In [411]:
# embark_town의 829행의 데이터 출력
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object

In [412]:
# embark_town 열의 승선도시 중에서 가장 많이 출현한 값으로 치환하기
most_freq = df.embark_town.value_counts(dropna=False).idxmax() # idxmax : 인덱스 중 맥스 

In [413]:
df['embark_town'].fillna(most_freq, inplace = True)

In [414]:
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829    Southampton
Name: embark_town, dtype: object

### 앞뒤의 유사성 데이터로 값을 치환

In [415]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [416]:
df['embark_town'][825:831]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
830      Cherbourg
Name: embark_town, dtype: object

In [417]:
# embark_town 열의 NaN 값을 바로 앞에 있는 828행의 값으로 변경하기
df['embark_town'].fillna(method = 'bfill', inplace=True) # ffill = frontfill
df['embark_town'][825:831]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829      Cherbourg
830      Cherbourg
Name: embark_town, dtype: object

In [418]:
missing_count

False    891
Name: alone, dtype: int64

---
# 중복 데이터 처리
- 동일한 대상이 중복으로 존재하는 것은 분석 결과의 왜곡을 초래한다.


In [419]:
import pandas as pd

In [420]:
df = pd.DataFrame({'c1':['a', 'a', 'b', 'a', 'b','a'],
                  'c2':[1, 1, 1, 2, 2,1],
                  'c3':[1, 1, 2, 2, 2,1]})

In [421]:
df

c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2
5  a   1   1

In [422]:
# 데이터프레임 전체 행 데이터 중에서 중복값 찾기
df.duplicated()

0    False
1     True
2    False
3    False
4    False
5     True
dtype: bool

In [423]:
# 특정 열에서 데이터 중복값 찾기
df['c2'].duplicated()

0    False
1     True
2     True
3    False
4     True
5     True
Name: c2, dtype: bool

### 중복데이터 제거

In [424]:
df.duplicated()
df.drop_duplicates()

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [425]:
# c2와 c3열을 기준으로 중복 행을 제거
df[['c2','c3']].duplicated()
df[['c2','c3']].drop_duplicates()
df.drop_duplicates(subset=['c2','c3'])

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2

---
# 데이터 표준화
- 여러 곳에서 수집한 데이터는 서로 다른 단위가 섞여 있거나 대상을 다른 형식으로 표현하는 경우가 많다.
- 대상을 표현하는 방법에 차이가 있으면 분석의 정확도는 현저히 낮아진다.
- 데이터 포맷을 일관성 있게 하는 작업이 표준화 이다.

In [426]:
df = pd.read_csv('../Data/auto-mpg.csv')
df.columns = ['mpg','cylinders','displacement','horsepower','weight',
              'acceleration','model year','origin','name']
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [427]:
df.to_csv("../Data/auto-mpg.csv", index= False)

# 단위 변환
mpg(mile per gallon) -> kpl(kilometer per liter) 로 변환    
1 mile = 1.60934 km   
1 gallon = 3.78541 liter  

In [428]:
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [429]:
df['kpl'] = df.mpg * 1.60934 /3.78541
df.kpl = df.kpl.round(2)
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   

   origin                       name   kpl  
0       1  chevrolet chevelle malibu  7.65  
1       1          buick skylark 320  6.38  
2       1         plymouth satellite  7.65  
3       1              amc rebel sst  6.80  
4       1                ford torino  7.23

### 자료형 변환

In [430]:
# 각 열의 자료형 확인
df.info()
df.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    float64
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   name          398 non-null    object 
 9   kpl           398 non-null    float64
dtypes: float64(5), int64(3), object(2)
memory usage: 31.2+ KB


mpg             float64
cylinders         int64
displacement    float64
horsepower       object
weight          float64
acceleration    float64
model year        int64
origin            int64
name             object
kpl             float64
dtype: object

> horsepower열이 object자료형이다. 엔진 출력을 나타내는 열이므로 숫자형으로 변환해 주는 것이 옳다.  
> 데이터를 보면 숫자처럼 보이는 데 자료형이 다른 것으로 나오는 것을 보면 데이터 중에 다른 자료가 있는 것이다.

In [431]:
df.horsepower.value_counts()['?']

6

In [432]:
# horsepower열의 고유값 확인
df['horsepower'].unique()


array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', '?', '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

In [433]:
# Data 갯수 확인
len(df.horsepower)

398

In [434]:
# Data가 ?인 것을 확인
mask = df['horsepower'] == '?'
mask

0      False
1      False
2      False
3      False
4      False
       ...  
393    False
394    False
395    False
396    False
397    False
Name: horsepower, Length: 398, dtype: bool

In [435]:
df[mask]

mpg  cylinders  displacement horsepower  weight  acceleration  \
32   25.0          4          98.0          ?  2046.0          19.0   
126  21.0          6         200.0          ?  2875.0          17.0   
330  40.9          4          85.0          ?  1835.0          17.3   
336  23.6          4         140.0          ?  2905.0          14.3   
354  34.5          4         100.0          ?  2320.0          15.8   
374  23.0          4         151.0          ?  3035.0          20.5   

     model year  origin                  name    kpl  
32           71       1            ford pinto  10.63  
126          74       1         ford maverick   8.93  
330          80       2  renault lecar deluxe  17.39  
336          80       1    ford mustang cobra  10.03  
354          81       2           renault 18i  14.67  
374          82       1        amc concord dl   9.78

In [436]:
# 누락 데이터(?)를 삭제 
# ? np.nan으로 변경
import numpy as np


In [437]:
df['horsepower'].replace('?',np.nan,inplace=True)
df['horsepower'].unique()

array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', nan, '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

In [438]:
# 누락 데이터 행을 삭제
df.dropna(subset=['horsepower'],axis=0, inplace=True)
len(df['horsepower'])

392

In [439]:
# horsepower type을 실수형으로 변경
df['horsepower'] = df['horsepower'].astype('float')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    float64
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   name          392 non-null    object 
 9   kpl           392 non-null    float64
dtypes: float64(6), int64(3), object(1)
memory usage: 33.7+ KB


In [440]:
# origin의 데이터를 문자형 데이터로 변형 (1:USA, 2:EU, 3:Japan)
# df.origin.replace([1,2,3],['USA','EU',"Japan"],inplace=True) 리스트 형태도 가능 
df.origin.replace({1:'USA',2:'EU',3:"Japan"},inplace=True) # 딕셔너리 형태도 가능

df.origin

0      USA
1      USA
2      USA
3      USA
4      USA
      ... 
393    USA
394     EU
395    USA
396    USA
397    USA
Name: origin, Length: 392, dtype: object

In [441]:
# origin 옆의 문자열 자료형을 범주형으로 변환
df['origin'] = df['origin'].astype('category')
df['origin'].dtypes

CategoricalDtype(categories=['EU', 'Japan', 'USA'], ordered=False)

In [442]:
# 범주형을 문자형으로 변환
df['origin'] = df['origin'].astype('str')
df['origin'].dtypes
# dtype('O') 는 Object

dtype('O')

In [443]:
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0  3504.0          12.0   
1  15.0          8         350.0       165.0  3693.0          11.5   
2  18.0          8         318.0       150.0  3436.0          11.0   
3  16.0          8         304.0       150.0  3433.0          12.0   
4  17.0          8         302.0       140.0  3449.0          10.5   

   model year origin                       name   kpl  
0          70    USA  chevrolet chevelle malibu  7.65  
1          70    USA          buick skylark 320  6.38  
2          70    USA         plymouth satellite  7.65  
3          70    USA              amc rebel sst  6.80  
4          70    USA                ford torino  7.23

In [444]:
# model year가 정수로 되어 있으나 수치계산과 관련 없으므로 범주형으로 변환한다.

df['model year'].sample(3)

253    78
397    82
39     71
Name: model year, dtype: int64

In [445]:
# model year 를 범주형으로 변환
df['model year'] = df['model year'].astype('category')
df['model year'].sample(3)

386    82
219    77
242    77
Name: model year, dtype: category
Categories (13, int64): [70, 71, 72, 73, ..., 79, 80, 81, 82]

---
### 범주형(카테고리) 데이터 처리
- 연속 데이터를 그대로 사용하기 보다는 일정한 구간(binning)으로 나눠서 분석하는 것이 효율적인 경우가 많다.

In [446]:
# horsepower를 3개의 범위로 나누기

count, bin_dividers = np.histogram(df['horsepower'], bins = 3)
np.histogram(df['horsepower'], bins = 3)

(array([257, 103,  32]),
 array([ 46.        , 107.33333333, 168.66666667, 230.        ]))

> 46 ~ 107.33333333 , 107.33333333 ~ 168.66666667, 168.66666667 ~ 230

In [447]:
# 3 개의 bin에 이름 지정
bin_names = ['저출력','보통출력','고출력']


In [448]:
df['hp_bin'] = pd.cut(
    x=df['horsepower'],
    bins= bin_dividers,
    labels=bin_names,
    include_lowest= True # 첫 경계값이 왼쪽 구간에 포함 107.3333 이 첫 구간에 속함
)

In [449]:
df[['horsepower','hp_bin']].head(15)

horsepower hp_bin
0        130.0   보통출력
1        165.0   보통출력
2        150.0   보통출력
3        150.0   보통출력
4        140.0   보통출력
5        198.0    고출력
6        220.0    고출력
7        215.0    고출력
8        225.0    고출력
9        190.0    고출력
10       170.0    고출력
11       160.0   보통출력
12       150.0   보통출력
13       225.0    고출력
14        95.0    저출력


---
### 더미(Dummy) 변수 
- 범주형 데이터를 회귀분석이나 머신러닝에 바로 사용할 수 없는 경우 사용.
- 0 또는 1로 표현

In [450]:
horsepower_dummies = pd.get_dummies(df['hp_bin'])
horsepower_dummies.head()

저출력  보통출력  고출력
0    0     1    0
1    0     1    0
2    0     1    0
3    0     1    0
4    0     1    0

--- 
### One-Hot Encoding
- 범주형 데이터를 0,1을 갖는 원핫벡터로 변환     
- 결과는 선형대수학에서 정의하는 희소행렬(sparse matrix)로 변환한다     
- 희소행렬이란 ? (행,열) 좌표와 값 형태로 정리     
- sklearn 모듈을 이용 

In [451]:
from sklearn import preprocessing

In [452]:
# 전처리를 위한 encoder 객체 만들기 
label_encoder = preprocessing.LabelEncoder() #label_encoder 생성 
onehot_encoder = preprocessing.OneHotEncoder() #onehot_encoder 생성 

In [453]:
# label encoder로 문자열 범주 -> 숫자형 범주로 변환 
# 3종류니까 0,1,2 
onehot_labeled= label_encoder.fit_transform(df['hp_bin'].head(15))
onehot_labeled

array([1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 2])

In [454]:
# 2차원 행렬로 형태 변경 (희소행렬로 보여주기 위해)
onehot_reshaped = onehot_labeled.reshape(len(onehot_labeled), 1) # 15행 1열 # 2차원 배열 
onehot_reshaped

array([[1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [2]])

In [455]:
# 희소행렬로 변환 (좌표 - 결과값)
onehot_fitted = onehot_encoder.fit_transform(onehot_reshaped)
print(onehot_fitted)

# (14,2) 1.0 : 14행, 2번이 1이니까 고출력 
# --> 이렇게 하면 컬럼 3개 안만들어도 어디에 1 값인 줄 알 수 있음 

  (0, 1)	1.0
  (1, 1)	1.0
  (2, 1)	1.0
  (3, 1)	1.0
  (4, 1)	1.0
  (5, 0)	1.0
  (6, 0)	1.0
  (7, 0)	1.0
  (8, 0)	1.0
  (9, 0)	1.0
  (10, 0)	1.0
  (11, 1)	1.0
  (12, 1)	1.0
  (13, 0)	1.0
  (14, 2)	1.0


--- 
### 정규화(Normalization)
- 각 열에 있는 숫자 데이터의 상대적 크기 때문에 머신러닝 분석결과가 달라질 수 있음     
- 숫자 데이터의 상대적인 크기 차이를 제거 (0,1 쪽으로 데이터를 몰아넣음)    
- 최대값, 평균값, 절대값 등을 이용하여 정규화 하는 방법들 있음    

In [456]:
# 최대값의 절대값을 이용한 정규화 
df.horsepower.describe()

count    392.000000
mean     104.469388
std       38.491160
min       46.000000
25%       75.000000
50%       93.500000
75%      126.000000
max      230.000000
Name: horsepower, dtype: float64

In [457]:
# 방법1) horsepower 열의 최대값의 절대값으로 모든 데이터를 나누기 하여 저장 
horsepower1 = df.horsepower / abs(df.horsepower.max())

In [458]:
horsepower1.describe()

count    392.000000
mean       0.454215
std        0.167353
min        0.200000
25%        0.326087
50%        0.406522
75%        0.547826
max        1.000000
Name: horsepower, dtype: float64

In [459]:
# 방법2) 각 열의 데이터 중에 최대값과 최소값을 뺄셈한 값으로 나눠 저장하는 방법 
# - 각 열에서 해당 열의 최소값을 뺀 값을 분자로 하고 
min_x = df.horsepower - df.horsepower.min()
# - 해당 열의 최대, 최소값의 차르 분모로 하는 수를 계산하면 가장 큰 값은 1이 된다 
min_max = df.horsepower.max() - df.horsepower.min()
horsepower2 = min_x / min_max
horsepower2.describe()

count    392.000000
mean       0.317768
std        0.209191
min        0.000000
25%        0.157609
50%        0.258152
75%        0.434783
max        1.000000
Name: horsepower, dtype: float64

---
# 시계열 데이터
- 특정한 시점을 기록하는 Timestamp
- 두 지점 사이의 일정한 기간을 나타내는 Period

In [460]:
df = pd.read_csv('../Data/stock-data.csv')
df.head()

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

In [461]:
df.isnull().sum()

Date      0
Close     0
Start     0
High      0
Low       0
Volume    0
dtype: int64

In [462]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Date    20 non-null     object
 1   Close   20 non-null     int64 
 2   Start   20 non-null     int64 
 3   High    20 non-null     int64 
 4   Low     20 non-null     int64 
 5   Volume  20 non-null     int64 
dtypes: int64(5), object(1)
memory usage: 1.1+ KB


In [463]:
df['new_Date'] = pd.to_datetime(df['Date'])
df.head()

Date  Close  Start   High    Low  Volume   new_Date
0  2018-07-02  10100  10850  10900  10000  137977 2018-07-02
1  2018-06-29  10700  10550  10900   9990  170253 2018-06-29
2  2018-06-28  10400  10900  10950  10150  155769 2018-06-28
3  2018-06-27  10900  10800  11050  10500  133548 2018-06-27
4  2018-06-26  10800  10900  11000  10700   63039 2018-06-26

In [464]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Date      20 non-null     object        
 1   Close     20 non-null     int64         
 2   Start     20 non-null     int64         
 3   High      20 non-null     int64         
 4   Low       20 non-null     int64         
 5   Volume    20 non-null     int64         
 6   new_Date  20 non-null     datetime64[ns]
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 1.2+ KB


In [465]:
df.set_index('new_Date',inplace=True)
df.head()

Date  Close  Start   High    Low  Volume
new_Date                                                  
2018-07-02  2018-07-02  10100  10850  10900  10000  137977
2018-06-29  2018-06-29  10700  10550  10900   9990  170253
2018-06-28  2018-06-28  10400  10900  10950  10150  155769
2018-06-27  2018-06-27  10900  10800  11050  10500  133548
2018-06-26  2018-06-26  10800  10900  11000  10700   63039

In [466]:
df.drop('Date',axis=1,inplace=True)
df.head()

Close  Start   High    Low  Volume
new_Date                                      
2018-07-02  10100  10850  10900  10000  137977
2018-06-29  10700  10550  10900   9990  170253
2018-06-28  10400  10900  10950  10150  155769
2018-06-27  10900  10800  11050  10500  133548
2018-06-26  10800  10900  11000  10700   63039

In [467]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 20 entries, 2018-07-02 to 2018-06-01
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Close   20 non-null     int64
 1   Start   20 non-null     int64
 2   High    20 non-null     int64
 3   Low     20 non-null     int64
 4   Volume  20 non-null     int64
dtypes: int64(5)
memory usage: 960.0 bytes


### Timestamp를 Period로 변환

In [468]:
# 날짜 형식의 문자열로 구성되는 리스트 정의
dates = ['2019-01-01', '2020-03-01', '2021-06-01']

In [469]:
ts_dates = pd.to_datetime(dates)
ts_dates

DatetimeIndex(['2019-01-01', '2020-03-01', '2021-06-01'], dtype='datetime64[ns]', freq=None)

In [470]:
pr_day = ts_dates.to_period(freq = 'D')
print(pr_day)
pr_day = ts_dates.to_period(freq = 'M')
print(pr_day)
pr_day = ts_dates.to_period(freq = 'A') # A 대신 Y 가능
print(pr_day)

PeriodIndex(['2019-01-01', '2020-03-01', '2021-06-01'], dtype='period[D]')
PeriodIndex(['2019-01', '2020-03', '2021-06'], dtype='period[M]')
PeriodIndex(['2019', '2020', '2021'], dtype='period[A-DEC]')


> freq Option   
> D (Day: 1일)   
> W (Week:1주)   
> M (Month end: 월말)   
> MS (Month begin: 월초)   
> Q (Quarter end : 분기말)   
> QS (Quarter begin: 분기초)   
> A (year end : 연말)   
> AS (year begin: 연초)   
> B (Business day : 휴일 제외)   
> H (hour : 1시간)   
> T (minute :1분)   
> S (second: 1초)

### 시계열 데이터 만들기

In [471]:
# Timestamp의 배열 만들기 : 월 간격
ts_ms = pd.date_range(
    start = '2022-01-01', # 날짜 범위의 시작
    end = None, # 날짜 범위의 끝
    periods = 6, # 생성할 Timestamp 의 갯수
    freq = 'MS', # 시간 간격(MS: 월의 시작일),
    tz = 'Asia/Seoul' # timezone 시간대 
)
ts_ms

DatetimeIndex(['2022-01-01 00:00:00+09:00', '2022-02-01 00:00:00+09:00',
               '2022-03-01 00:00:00+09:00', '2022-04-01 00:00:00+09:00',
               '2022-05-01 00:00:00+09:00', '2022-06-01 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='MS')

In [472]:
# 월 간격, 월의 마지막 날 기준
ts_me = pd.date_range(
    start = '2022-01-01', # 날짜 범위의 시작
    end = None, # 날짜 범위의 끝
    periods = 6, # 생성할 Timestamp 의 갯수
    freq = 'M', # 시간 간격(MS: 월의 마지막날),
    tz = 'Asia/Seoul' # timezone 시간대 
)
ts_me

DatetimeIndex(['2022-01-31 00:00:00+09:00', '2022-02-28 00:00:00+09:00',
               '2022-03-31 00:00:00+09:00', '2022-04-30 00:00:00+09:00',
               '2022-05-31 00:00:00+09:00', '2022-06-30 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='M')

In [473]:
# 분기(3개월) 간격, 월의 마지막 날 기준
ts_3m = pd.date_range(
    start= '2022-01-01',
    end= None,
    periods= 6,
    freq = '3M', # 시간 간격 계수 붙이기 가능
    tz = 'Asia/Seoul'
)
ts_3m


DatetimeIndex(['2022-01-31 00:00:00+09:00', '2022-04-30 00:00:00+09:00',
               '2022-07-31 00:00:00+09:00', '2022-10-31 00:00:00+09:00',
               '2023-01-31 00:00:00+09:00', '2023-04-30 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='3M')

---
### Period(기간) 배열


In [474]:
# 1개월 길이
pr_m = pd.period_range(
    start = '2022-01-01', # 날짜 범위의 시작
    end = None, # 날짜 범위의 끝
    periods=3, # 생성할 Period 개수
    freq = 'M'
)
pr_m


PeriodIndex(['2022-01', '2022-02', '2022-03'], dtype='period[M]')

In [475]:
# 1시간 길이
pr_h = pd.period_range(
    start = '2022-01-01', # 날짜 범위의 시작
    end = None, # 날짜 범위의 끝
    periods=3, # 생성할 Period 개수
    freq = 'H'
)
pr_h



PeriodIndex(['2022-01-01 00:00', '2022-01-01 01:00', '2022-01-01 02:00'], dtype='period[H]')

In [476]:
# 2시간 길이
pr_2h = pd.period_range(
    start = '2022-01-01', # 날짜 범위의 시작
    end = None, # 날짜 범위의 끝
    periods=3, # 생성할 Period 개수
    freq = '2H'
)
pr_2h



PeriodIndex(['2022-01-01 00:00', '2022-01-01 02:00', '2022-01-01 04:00'], dtype='period[2H]')

In [477]:
df.reset_index(inplace=True)
df.head()

new_Date  Close  Start   High    Low  Volume
0 2018-07-02  10100  10850  10900  10000  137977
1 2018-06-29  10700  10550  10900   9990  170253
2 2018-06-28  10400  10900  10950  10150  155769
3 2018-06-27  10900  10800  11050  10500  133548
4 2018-06-26  10800  10900  11000  10700   63039

In [478]:
df['Year'] = df['new_Date'].dt.year
df['Month'] = df['new_Date'].dt.month
df['Day'] = df['new_Date'].dt.day
df.head()

new_Date  Close  Start   High    Low  Volume  Year  Month  Day
0 2018-07-02  10100  10850  10900  10000  137977  2018      7    2
1 2018-06-29  10700  10550  10900   9990  170253  2018      6   29
2 2018-06-28  10400  10900  10950  10150  155769  2018      6   28
3 2018-06-27  10900  10800  11050  10500  133548  2018      6   27
4 2018-06-26  10800  10900  11000  10700   63039  2018      6   26

In [479]:
df['Date_yr'] = df['new_Date'].dt.to_period(freq='A')
df['Date_m'] = df['new_Date'].dt.to_period(freq='M')
df.head()

new_Date  Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
0 2018-07-02  10100  10850  10900  10000  137977  2018      7    2    2018   
1 2018-06-29  10700  10550  10900   9990  170253  2018      6   29    2018   
2 2018-06-28  10400  10900  10950  10150  155769  2018      6   28    2018   
3 2018-06-27  10900  10800  11050  10500  133548  2018      6   27    2018   
4 2018-06-26  10800  10900  11000  10700   63039  2018      6   26    2018   

    Date_m  
0  2018-07  
1  2018-06  
2  2018-06  
3  2018-06  
4  2018-06

In [480]:
df.set_index('new_Date',inplace=True)
df.head()

Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
new_Date                                                                   
2018-07-02  10100  10850  10900  10000  137977  2018      7    2    2018   
2018-06-29  10700  10550  10900   9990  170253  2018      6   29    2018   
2018-06-28  10400  10900  10950  10150  155769  2018      6   28    2018   
2018-06-27  10900  10800  11050  10500  133548  2018      6   27    2018   
2018-06-26  10800  10900  11000  10700   63039  2018      6   26    2018   

             Date_m  
new_Date             
2018-07-02  2018-07  
2018-06-29  2018-06  
2018-06-28  2018-06  
2018-06-27  2018-06  
2018-06-26  2018-06

In [481]:
#날짜 인덱스를 이용하여 데이터 선택하기
df_y = df.loc['2018']
df_y.head(2)

Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
new_Date                                                                   
2018-07-02  10100  10850  10900  10000  137977  2018      7    2    2018   
2018-06-29  10700  10550  10900   9990  170253  2018      6   29    2018   

             Date_m  
new_Date             
2018-07-02  2018-07  
2018-06-29  2018-06

In [482]:
df_y = df.loc['2018-07','Start':'High']
df_y.head()

Start   High
new_Date                
2018-07-02  10850  10900

In [483]:
df.head()

Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
new_Date                                                                   
2018-07-02  10100  10850  10900  10000  137977  2018      7    2    2018   
2018-06-29  10700  10550  10900   9990  170253  2018      6   29    2018   
2018-06-28  10400  10900  10950  10150  155769  2018      6   28    2018   
2018-06-27  10900  10800  11050  10500  133548  2018      6   27    2018   
2018-06-26  10800  10900  11000  10700   63039  2018      6   26    2018   

             Date_m  
new_Date             
2018-07-02  2018-07  
2018-06-29  2018-06  
2018-06-28  2018-06  
2018-06-27  2018-06  
2018-06-26  2018-06

In [484]:
# 시간 간격 계산 . 최근 180일 ~ 189일 사이의 값들만 선택하기
today = pd.to_datetime('2018-12-25')
df['time_delta'] = today - df.index
df.set_index('time_delta',inplace=True)
df_180 = df['180 days': '189 days']
df_180

Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
time_delta                                                                 
180 days    10400  10900  10950  10150  155769  2018      6   28    2018   
181 days    10900  10800  11050  10500  133548  2018      6   27    2018   
182 days    10800  10900  11000  10700   63039  2018      6   26    2018   
183 days    11150  11400  11450  11000   55519  2018      6   25    2018   
186 days    11300  11250  11450  10750  134805  2018      6   22    2018   
187 days    11200  11350  11750  11200  133002  2018      6   21    2018   
188 days    11550  11200  11600  10900  308596  2018      6   20    2018   
189 days    11300  11850  11950  11300  180656  2018      6   19    2018   

             Date_m  
time_delta           
180 days    2018-06  
181 days    2018-06  
182 days    2018-06  
183 days    2018-06  
186 days    2018-06  
187 days    2018-06  
188 days    2018-06  
189 days    2018-06

In [485]:
df_180.loc['180 days':'190 days']

Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
time_delta                                                                 
180 days    10400  10900  10950  10150  155769  2018      6   28    2018   
181 days    10900  10800  11050  10500  133548  2018      6   27    2018   
182 days    10800  10900  11000  10700   63039  2018      6   26    2018   
183 days    11150  11400  11450  11000   55519  2018      6   25    2018   
186 days    11300  11250  11450  10750  134805  2018      6   22    2018   
187 days    11200  11350  11750  11200  133002  2018      6   21    2018   
188 days    11550  11200  11600  10900  308596  2018      6   20    2018   
189 days    11300  11850  11950  11300  180656  2018      6   19    2018   

             Date_m  
time_delta           
180 days    2018-06  
181 days    2018-06  
182 days    2018-06  
183 days    2018-06  
186 days    2018-06  
187 days    2018-06  
188 days    2018-06  
189 days    2018-06

In [486]:
## Batch : 일괄, 배치 처리는 일괄 처리